### PR TITLE
feat: add batch functions and tutorial

### PR DESCRIPTION
**Pull request by jpetot**

### CODE REVIEW

File: notebooks/Tutorial7-Batch_avatarization.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+#       format_name: light
+#       format_version: '1.5'
+#       jupytext_version: 1.14.5
+# ---
+
+# # Tutorial 7: Batch avatarization
+
+# In this tutorial, we will perform the avatarization on batch of data. This can be useful if you have to much data that can be avatarized in one shot.
+#
+# # TODO: add a schema of the process
```
[the remark this answers is not in the document]
Comment (jpetot) [reply notebooks/Tutorial7-Batch_avatarization.py:15]: you are right, it's now done

File: avatars/api_test.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    )
+
+    order = {
+        batch_result.training_result.original_id: pd.Index([0, 2]),
+        batch_result.batch_results[0].original_id: pd.Index([3, 1]),
+    }
+
+    client = Mock()
+    client.pandas_integration.download_dataframe.side_effect = [train, split]
```
[the remark this answers is not in the document]
Comment (tcrasset) [reply avatars/api_test.py:60]: The following isn't really the scope of this PR, but we should take most of the stuff that was setup for `integration_test` in the `avatar` repo and copy it here, e.g. the `client` fixture in `conftest.py` that was used to authenticate the client, and mock it, etc..

Then, you wouldn't have to mock the whole `client` but only `download_dataframe` with `unittest.mock.patch`

File: avatars/lib/split.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+) -> Tuple[pd.DataFrame, List[pd.DataFrame]]:
+    """Create batches from a dataframe with the format (training, splits).
+
+
+    Arguments
+    ---------
+        df:
+            dataframe to split into batch
+        row_limit:
+            maximum number of row by batch, this parameter determine the number of batch that will be created
```
Comment (tcrasset) [inline avatars/lib/split.py:21]:
```suggestion
            maximum number of rows per batch, this parameter determines the number of batches that will be created
```

File: avatars/lib/split.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    -----------------
+        seed:
+            seed of the random split for the batch
+
+    Returns
+    -------
+        training:
+            the training batch that will be used to fit the anonymization
+        training:
+            splits: all other batches
```
Comment (tcrasset) [inline avatars/lib/split.py:33]:
```suggestion

        splits: all other batches
```

File: avatars/lib/split.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    >>> splits
+    [    a  b
+    9   1  a
+    4   3  b
+    11  5  a
+    6   1  a
+    5   5  a
+    8   5  a]
+    """
+    number_of_split = math.ceil(len(df) / row_limit)
```
Comment (tcrasset) [inline avatars/lib/split.py:61]:
```suggestion
    number_of_splits = math.ceil(len(df) / row_limit)
```
nit

File: avatars/lib/split.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    categorical_columns = df.select_dtypes(include=["object", "category"]).columns
+
+    index = get_index_without_duplicated_modalities(df[categorical_columns])
+    first_occurrence = df.loc[index, :]
+    if len(first_occurrence) > row_limit * 0.5:
+        raise ValueError(
+            "You have too much variability in your categorical data."
+            "You can increase your `row_limit`, reduce your categorical variability, or built "
+            "your own batch pipeline."
```
Comment (tcrasset) [inline avatars/lib/split.py:70]:
```suggestion
            "You can increase your `row_limit`, reduce your categorical variability, or build "
            "your own batch pipeline."
```

File: avatars/lib/split.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    all_df = pd.concat([first_occurrence, shuffled])
+    splits = np.array_split(all_df, number_of_split)
+    splits = [pd.DataFrame(split) for split in splits]
+    training = splits.pop(0)
+
+    return training, splits
+
+
+def get_index_without_duplicated_modalities(df: pd.DataFrame) -> List[int]:
+    values: Dict[str, Dict[str, bool]] = {col: {} for col in df.columns}
```
Comment (tcrasset) [inline avatars/lib/split.py:86]: docstring

File: avatars/lib/split.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    df_array = np.array(df)
```
Comment (tcrasset) [inline avatars/lib/split.py:87]: You are duplicating memory. Use `df.values` or `df.to_numpy()`

File: avatars/lib/split.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    index_to_keep = []
+    for i in range(df_array.shape[0]):
+        should_keep_row = False
+        for j in range(df_array.shape[1]):
+            try:
+                values[df.columns[j]][df_array[i, j]]
+            except KeyError:
+                values[df.columns[j]][df_array[i, j]] = True
+                should_keep_row = True
```
Comment (tcrasset) [inline avatars/lib/split.py:96]: Small comment explaining what is done?

Comment (tcrasset) [inline avatars/lib/split.py:96]:
```suggestion
                values[df.columns[j]][df_array[i, j]] = True
                should_keep_row = True
                break
```

Comment (tcrasset) [reply avatars/lib/split.py:96]: Once you find the first unique value, you can break and add the row.

File: notebooks/Tutorial7-Batch_avatarization.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+#
+#  ![image.png](attachment:image.png)
+#
+
+# ## Connection
+
+# +
+import warnings
+
+warnings.filterwarnings("ignore")
```
Comment (tcrasset) [inline notebooks/Tutorial7-Batch_avatarization.py:28]: Why?

We should fix the warnings.

Comment (tcrasset) [reply notebooks/Tutorial7-Batch_avatarization.py:28]: This shouldn't be in public documentation, shows bad practice from our part

File: notebooks/Tutorial7-Batch_avatarization.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+# Verify that we can connect to the API server
+client.health.get_health()
+# -
+
+# ## Load the data
+# We will use a subset of the dataset `adult_with_missing`.
+
+df = pd.read_csv("../fixtures/adult_with_missing.csv").iloc[:1000, :]
+
```
Comment (tcrasset) [inline notebooks/Tutorial7-Batch_avatarization.py:101]:
```suggestion
df = pd.read_csv("../fixtures/adult_with_missing.csv").head(1000)

```

File: notebooks/Tutorial7-Batch_avatarization.py [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+# +
+# create some batches with from the df
+
+RowLimit = 200
```
Comment (tcrasset) [inline notebooks/Tutorial7-Batch_avatarization.py:105]:
```suggestion
row_limit = 200
```
`snake_case` for python variables

File: notebooks/Tutorial7-Batch_avatarization.py [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+    df,
+    row_limit=RowLimit,
+)
+# -
+
+# ## Launch batch avatarization
+#
+
+# +
+dataset_ref_id, dataset_splited_ids, order = upload_batch_and_get_order(
```
Comment (tcrasset) [inline notebooks/Tutorial7-Batch_avatarization.py:117]:
```suggestion
dataset_training_id, dataset_splited_ids, order = upload_batch_and_get_order(
```

Comment (tcrasset) [inline notebooks/Tutorial7-Batch_avatarization.py:117]:
```suggestion
dataset_ref_id, dataset_split_ids, order = upload_batch_and_get_order(
```
same as above

File: notebooks/Tutorial7-Batch_avatarization.py [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+            dataset_ids=dataset_splited_ids,
+            k=20,
+            imputation=ImputationParameters(method=ImputeMethod.mean),
+        )
+    )
+)
+batch_job = client.jobs.get_avatarization_batch_job(batch_job.id, timeout=10)
+# -
+
+batch_job = client.jobs.get_avatarization_batch_job(batch_job.id, timeout=10000)
```
Comment (tcrasset) [inline notebooks/Tutorial7-Batch_avatarization.py:135]:
```suggestion
# -

batch_job = client.jobs.get_avatarization_batch_job(batch_job.id, timeout=10000)
```
nit: only one get

File: notebooks/Tutorial7-Batch_avatarization.py [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+    )
+)
+privacy_job = client.jobs.get_privacy_metrics_batch_job(
+    privacy_job_ref.id, timeout=100000
+)
+# you can access batch-averaged metrics
+print("Mean metrics")
+print(privacy_job.result.mean_metrics)
+
+# you can access to worst metric over all batches
```
Comment (tcrasset) [inline notebooks/Tutorial7-Batch_avatarization.py:158]:
```suggestion
# you can access the worst metrics over all batches
```

File: notebooks/Tutorial7-Batch_avatarization.py [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+        ),
+    )
+)
+signal_job = client.jobs.get_signal_metrics_batch_job(signal_job_ref.id)
+
+print("Mean metrics")
+print(signal_job.result.mean_metrics)
+# -
+
+# ## Built the anonymized dataset
```
Comment (tcrasset) [inline notebooks/Tutorial7-Batch_avatarization.py:181]:
```suggestion
# ## Build the anonymized dataset
```

File: notebooks/Tutorial7-Batch_avatarization.py [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+# you can access to worst metric over all batches
+print("Worst metrics")
+print(privacy_job.result.worst_metrics)
+# -
+
+# ## Launch signal metrics per batch
+#
+
+# +
+signal_job_ref = client.jobs.create_signal_metrics_batch_job(
```
Comment (tcrasset) [inline notebooks/Tutorial7-Batch_avatarization.py:167]:
```suggestion
signal_job_training = client.jobs.create_signal_metrics_batch_job(
```

File: notebooks/Tutorial7-Batch_avatarization.py [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+
+print("Mean metrics")
+print(signal_job.result.mean_metrics)
+# -
+
+# ## Built the anonymized dataset
+#
+#
+
+# ### Get the shuflle avatar dataframe
```
Comment (tcrasset) [inline notebooks/Tutorial7-Batch_avatarization.py:185]:
```suggestion
# ### Get the shuffled avatar dataframe
```

File: notebooks/Tutorial7-Batch_avatarization.py [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+
+avatars = download_avatar_dataset_from_batch_result(batch_job.result, client=client)
+avatars.head()
+
+# ###  Get the sensitive unshuffle avatar dataframe
```
Comment (tcrasset) [inline notebooks/Tutorial7-Batch_avatarization.py:190]:
```suggestion
# ###  Get the sensitive unshuffled avatar dataframe
```

**Review by tcrasset — CHANGES_REQUESTED** [review]

File: avatars/api.py [hunk 20]
```diff
@@ -1037,4 +1191,109 @@ def avatarization_pipeline_with_processors(
...
+    for dataset, dataframe in zip(datasets_splited_id, splits):
+        batch_order[dataset] = dataframe.index
+
+    return training_dataset.id, datasets_splited_id, batch_order
+
+
+def download_avatar_dataset_from_batch_result(
+    avatarization_batch_result: AvatarizationBatchResult, client: "ApiClient"
+) -> pd.DataFrame:
+    """Download shuffled avatar from batch result.
```
[the remark this answers is not in the document]
Comment (tcrasset) [reply avatars/api.py:1234]: *the shuffled

File: avatars/api.py [hunk 21]
```diff
@@ -1037,4 +1194,119 @@ def avatarization_pipeline_with_processors(
         )
 
 
+def upload_batch_and_get_order(
+    client: "ApiClient",
+    training: pd.DataFrame,
+    splits: List[pd.DataFrame],
+    timeout: int = 10,
```
Comment (tcrasset) [inline avatars/api.py:1201]:
```suggestion
    timeout: int = DEFAULT_TIMEOUT,
```
I think there is a variable called like this, either in `api.py` or `client.py`

File: avatars/api.py [hunk 22]
```diff
@@ -1037,4 +1194,119 @@ def avatarization_pipeline_with_processors(
...
+) -> Tuple[UUID, List[UUID], Dict[UUID, pd.Index]]:
+    """Upload batches to the server
+    Arguments
+    ---------
+        training:
+            Dataframe used to train the anonymization model. This dataframe should contain all modalities of the categorical variables.
+        splits:
+            All other batches
+        client:
+            Api client
```
Comment (tcrasset) [inline avatars/api.py:1211]: Arguments needs reordering to match the function definition

File: avatars/api_test.py [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+            avatars_dataset=Dataset(
+                id=uuid4(), hash="plouf", download_url="truc", nb_dimensions=1
+            ),
+            sensitive_unshuffled_avatars_datasets=Dataset(
+                id=uuid4(), hash="plouf", download_url="truc", nb_dimensions=1
+            ),
+            original_id=uuid4(),
+        )
+    ],
+)
```
[the remark this answers is not in the document]
Comment (tcrasset) [reply avatars/api_test.py:35]: I still think we should use pytest fixtures, there are no downsides

File: avatars/lib/split.py (hunk 6 above)
Comment (tcrasset) [reply avatars/lib/split.py:70]: still there, `built` is past participle, we want the present

File: doc/source/avatars/processors.DatetimeProcessor.rst [hunk 24]
```diff
@@ -5,4 +5,4 @@ processors.DatetimeProcessor
 ============================
 
 .. autoclass:: processors.DatetimeProcessor
-   :members:
+   :members:@
```
Comment (tcrasset) [inline doc/source/avatars/processors.DatetimeProcessor.rst:8]: Typo?

**Review by tcrasset — CHANGES_REQUESTED** [review]
Very close to merging, mainly want the `DEFAULT_TIMEOUT` to be used, and strongly advise using pytest fixtures.

**Pull request merged**